### ISSUE TITLE
Parts of a stream in ASCII85 not decoded by qpdf

### ISSUE THREAD

**Issue by rbrito** [qpdf/qpdf#278]
Dear @jberkenbilt,

I have been using qpdf as a sort of preprocessor for [pdfsizeopt](https://github.com/pts/pdfsizeopt/), so that pdfsizeopt can reduce further the size of PDF files.

While playing with both tools, I found a **very** large file that had, essentially, all pages with images first deflated and, then, encoded as ASCII85. In fact, the file is (in essence) one very large text file, with essentially no blobs.

I used the following command, expecting it to decode the ASCII85 stream, the streams still remained there:

```
qpdf --stream-data=uncompress --compress-streams=n --decode-level=specialized p1.pdf p1.unc.pdf
```

I am attaching the file here as a sample. Are those some missing cases of decoding?

Thanks,

Rogério Brito.

[p1.pdf](https://github.com/qpdf/qpdf/files/2789449/p1.pdf)

**Comment by jberkenbilt** [comment]
The part that's not being decoded is an inline image. There's a very old outstanding issue still from the sourceforge issue tracker about this exact issue, which is mentioned in the TODO list. I've updated the TODO list to mention this issue. At the time, qpdf couldn't have done this easily, but at this point, it may be an easy fix. I will put a "next" label on it and take a look to see if I can do it soon. No guarantees.

That said, qpdf definitely chokes on this file in a way I will have to investigate. It is reporting a lexical error that may not be present, and when I attempt to edit the file manually to recompress the inline image, the result is not viewable by atril (a.k.a. evince). The PDF spec recommends using inline images for very small images, not full page images, so the file is already suspect, but either way, qpdf should not choke on it. I will investigate and let you know what I find. Thanks for the report.

**Comment by jberkenbilt** [comment]
This is a really interesting PDF file. The string "EI" actually appears in the *compressed* image data and also in the base85 encoding of the compressed data, and not only that, EI appears at the beginning of a line followed by a delimiter in the base85 data. When I wrote the lexical code for inline images, I realized this was a possibility, but I had no way to actually construct a valid file that had these properties. I wish I could put this file in qpdf's test suite, but I avoid files that may be copyrighted. In any case, it shouldn't be difficult to fix qpdf's tokenizer to properly handle this case.

**Comment by jberkenbilt** [comment]
I would also point out that, in the case, the decoded ascii85 data is already compressed, so compressing the ascii85 data results in something comparable in size to the original compressed data. If I manually re-encode the data without ascii85 and re-compress the streams, the file actually slightly *increases* in size. It's likely that using JPEG compression would improve things significantly. If this were not an inline image, qpdf 8.3.0's new --optimize-images flag would do this. It would be interesting to make that work with inline images too.

**Comment by jberkenbilt** [comment]
Just for fun, I manually optimized this PDF by converting the inline image to a regular image (by hand in a text editor) and then running qpdf 8.3.0's optimize-images.

[p1-opt.pdf](https://github.com/qpdf/qpdf/files/2793186/p1-opt.pdf)

**Comment by jberkenbilt** [comment]
I think it would be great if qpdf could actually do this automatically by converting the inline image to a regular image.

**Comment by jberkenbilt** [comment]
I've studied this in depth and written up some notes on how to fix it, but it's too complicated for me to fix this round. I will get to it, but not for 8.4.

**Comment by jberkenbilt** [comment]
Note: leaving the bug label on this because qpdf is not correctly finding the correct end of inline image. Given the way the tokenizer works, this is not an easy fix. Most likely I will have to add a separate method to locate EI and then modify the tokenizer to accept a hint about where it is. This will have to be invoked from both places in qpdf that use the tokenizer. Details in TODO file.

**Comment by jberkenbilt** [comment]
Actually I think I am going to try to do this for 8.4.0.

**Comment by jberkenbilt** [comment]
My work branch now properly parses this sample file. I have constructed a file myself with images that are not subject to copyright (like my own photo) that exhibit exactly the same behavior for inclusion in the test suite. I am going to try to modify optimize images to handle this case. The simple act of uncompressing the stream itself will never remove the Ascii85 data because that data is part of the inline image. However a separate method will be provided to extract the inline image and replace it with a real image which can be more optimally compressed.

**Comment by jberkenbilt** [comment]
My work branch now successfully optimizes images in this file. This fix will be in 8.4.0.
